File: scripts/thread-sync/thread-sync.py
#!/usr/bin/env python3

import threading

lock = threading.Lock()
resource = 0

def worker(id):
    global resource

    with lock:
        resource += 1
        print(f"In thread {id}, shared {resource=}")

threads = []
for i in range(1, 6):
    t = threading.Thread(target=worker, args=(i,))
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print(f"Final value of shared {resource=}")

File: scripts/thread-sync/thread-sync.pl
#!/usr/bin/env perl

use v5.38;
use threads;
use threads::shared;
use Thread::Semaphore;

my $semaphore = Thread::Semaphore->new(1);
my $resource :shared = 0;

sub worker {
    my $id = shift;

    $semaphore->down();

    $resource++;
    say "In thread $id, shared resource=$resource";

    $semaphore->up();
}

threads->create(\&worker, $_)->join foreach 1..5;

say "Final value of shared resource=$resource";
